hypervisor/src/mshv/aarch64/gic/mod.rs

```diff
@@ -0,0 +1,124 @@
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+//
+// Copyright © 2025, Microsoft Corporation
+//
+use std::any::Any;
+
+use serde::{Deserialize, Serialize};
+
+use crate::arch::aarch64::gic::{GicState, Result, Vgic, VgicConfig};
+use crate::{CpuState, Vm};
+
+pub struct MshvGicV2M {
+    /// GIC distributor address
+    pub dist_addr: u64,
+
+    /// GIC distributor size
+    pub dist_size: u64,
+
+    /// GIC re-distributors address
+    pub redists_addr: u64,
+
+    /// GIC re-distributors size
+    pub redists_size: u64,
+
+    /// GITS translator address
+    pub gits_addr: u64,
+
+    /// GITS translator size
+    pub gits_size: u64,
+
+    /// Number of CPUs handled by the device
+    pub vcpu_count: u64,
+}
+
+#[derive(Clone, Default, Serialize, Deserialize)]
+pub struct MshvGicV2MState {}
+
+impl From<GicState> for MshvGicV2MState {
+    fn from(state: GicState) -> Self {
+        match state {
+            GicState::Mshv(state) => state,
+            /* Needed in case other hypervisors are enabled */
+            #[allow(unreachable_patterns)]
+            _ => panic!("GicState is not valid"),
+        }
+    }
+}
+
+impl From<MshvGicV2MState> for GicState {
+    fn from(state: MshvGicV2MState) -> Self {
+        GicState::Mshv(state)
+    }
+}
+
+impl MshvGicV2M {
+    /// Create a new GICv2m device
+    pub fn new(_vm: &dyn Vm, config: VgicConfig) -> Result<MshvGicV2M> {
+        let gic_device = MshvGicV2M {
+            dist_addr: config.dist_addr,
+            dist_size: config.dist_size,
+            redists_addr: config.redists_addr,
+            redists_size: config.redists_size,
+            gits_addr: config.msi_addr,
+            gits_size: config.msi_size,
+            vcpu_count: config.vcpu_count,
+        };
+        Ok(gic_device)
+    }
+}
+
+impl Vgic for MshvGicV2M {
+    fn fdt_compatibility(&self) -> &str {
+        "arm,gic-v3"
```

Comment:
Is this correct?

Reply:
Yes, mshv exposes a gicv3 but uses gicv2m-frame for MSI interrupt support.